openvdb_ax/openvdb_ax/ast/Scanners.h

```diff
@@ -0,0 +1,203 @@
+// SPDX-License-Identifier: MPL-2.0
+
+/// @file ast/Scanners.h
+///
+/// @authors Nick Avramoussis, Richard Jones
+///
+/// @brief  Retrieve intrinsic information from AX AST by performing
+///   various traversal algorithms.
+///
+
+#ifndef OPENVDB_AX_COMPILER_AST_SCANNERS_HAS_BEEN_INCLUDED
+#define OPENVDB_AX_COMPILER_AST_SCANNERS_HAS_BEEN_INCLUDED
+
+#include "AST.h"
+#include "Visitor.h"
+
+#include <string>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+
+namespace ax {
+namespace ast {
+
+/// @brief  Returns whether or not a given branch of an AST reads from or writes
+///   to a given attribute.
+///
+/// @param node  The AST to analyze
+/// @param name  the name of the attribute to search for
+/// @param type  the type of the attribute to search for. If UNKNOWN, any
+///              attribute with the given name is checked.
+///
+bool usesAttribute(const ast::Node& node,
+    const std::string& name,
+    const tokens::CoreType type = tokens::UNKNOWN);
+
+/// @brief  Returns whether or not a given branch of an AST writes to a given
+///   attribute.
+///
+/// @param node  The AST to analyze
+/// @param name  the name of the attribute to search for
+/// @param type  the type of the attribute to search for. If UNKNOWN, the first
+///              attribute encountered with the given name is checked.
+///
+bool writesToAttribute(const ast::Node& node,
+    const std::string& name,
+    const tokens::CoreType type = tokens::UNKNOWN);
+
+/// @brief  Returns whether or not a given branch of an AST calls a function
+///
+/// @param node  The AST to analyze
+/// @param name  the name of the function to search for
+///
+bool callsFunction(const ast::Node& node, const std::string& name);
+
+/// @brief todo
+void catalogueVariables(const ast::Node& node,
+        std::vector<const ast::Variable*>* readOnly,
+        std::vector<const ast::Variable*>* writeOnly,
+        std::vector<const ast::Variable*>* readWrite,
+        const bool locals = true,
+        const bool attributes = true);
+
+/// @brief  Parse all attributes into three unique vectors which represent how they
+///         are accessed within the syntax tree. Read only attributes are stored
+///         within the 'readOnly' container (for example @code int a=@a; @endcode),
+///         write only attributes in the 'writeOnly' container @code @a=1; @endcode
+///         and readWrite attributes in the 'readWrite' container @code @a+=1; @endcode
+/// @note   Note that the code generator is able to do this far more efficiently, however
+///         this provides simple front-end support for detecting these types of operations
+///
+/// @param node       The AST to analyze
+/// @param readOnly   The unique list of attributes which are only read from
+/// @param writeOnly  The unique list of attributes which are only written too
+/// @param readWrite  The unique list of attributes which both read from and written too
+///
```

Comment:
Yeah this will currently mark a as writable. No plan to change this behaviour, at least not from the default - and we're doing code evaluation then this may be better suited to the IR level with llvms value visitors than at the AST